### ISSUE TITLE
TIP: bring requires_grad back

### ISSUE THREAD

**Issue by chenyuxyz** [tinygrad/tinygrad#16829]
https://github.com/tinygrad/tinygrad/issues/16520 shows an existing cross schedule bug that, for a sink that overwrites a Tensor, another Tensor that depends on the overwritten Tensor might be corrupted if it's not co-realized.

The solution is that when we know a sink overwrites a Tensor, we schedule everything that depends on that Tensor before the overwrite happens, thus the order and dependency are correct.

Now consider a training step
```
data = Tensor([1.0])
param = Tensor([1.0])
train_step(data, param)  # this updates param
```
Currently, we have a lazy `data.grad` that exists, but no one reads it, and it will not be realized and replaced by something else in the next step.
Once we fix the cross schedule bug, because `data.grad` depends on `param`, which got updated in `train_step`, the schedule would pull `data.grad` as something it needs to schedule and realize.

Bring back `requires_grad` with default False solves this, that's an API that explicit say don't create grad.

Alternative could be `train_step(data.detach(), param)`, or explicit backward targets like `loss.backward(targets=opt.params)`. Both reads weird and no one would remember to write like that.
